artifactory/commands/transferfiles/manager.go

```diff
@@ -188,30 +189,41 @@ func (ptm *PollingTasksManager) stop() {
 }
 
 type producerConsumerDetails struct {
-	producerConsumer parallel.Runner
-	errorsQueue      *clientUtils.ErrorsQueue
+	chunkBuilderProducerConsumer  parallel.Runner
+	chunkUploaderProducerConsumer parallel.Runner
+	errorsQueue                   *clientUtils.ErrorsQueue
 }
 
 func initProducerConsumer() producerConsumerDetails {
-	producerConsumer := parallel.NewRunner(GetThreads(), tasksMaxCapacity, false)
+	chunkUploaderProducerConsumer := parallel.NewRunner(GetThreads(), tasksMaxCapacity, false)
+	chunkBuilderProducerConsumer := parallel.NewRunner(GetThreads(), tasksMaxCapacity, false)
 	errorsQueue := clientUtils.NewErrorsQueue(1)
 
 	return producerConsumerDetails{
-		producerConsumer: producerConsumer,
-		errorsQueue:      errorsQueue,
+		chunkUploaderProducerConsumer: chunkUploaderProducerConsumer,
+		chunkBuilderProducerConsumer:  chunkBuilderProducerConsumer,
+		errorsQueue:                   errorsQueue,
 	}
 }
 
 func runProducerConsumer(pcDetails producerConsumerDetails, runWaitGroup *sync.WaitGroup) (runnerErr error, executionErr error) {
-	runWaitGroup.Add(1)
+	runWaitGroup.Add(2)
+	// When the producer consumer is idle for assumeProducerConsumerDoneWhenIdleForSeconds (not tasks are being handled)
+	// the work is assumed to be done.
+	go func() {
+		defer runWaitGroup.Done()
+		runnerErr = pcDetails.chunkUploaderProducerConsumer.DoneWhenAllIdle(assumeProducerConsumerDoneWhenIdleForSeconds)
```

Comment:
From code design perspective, I think it's unsafe to have an error returned from a go routine.
Instead of returning `runnerErr ` from this function let's use a local `err` variable, and add it to the log as an error message.
What I mean is, let's replace - 
```
runnerErr = pcDetails.chunkUploaderProducerConsumer.DoneWhenAllIdle(assumeProducerConsumerDoneWhenIdleForSeconds)
```
with
```
err := pcDetails.chunkBuilderProducerConsumer.DoneWhenAllIdle(assumeProducerConsumerDoneWhenIdleForSeconds)
log.Error("pcDetails.chunkBuilderProducerConsumer.DoneWhenAllIdle API failed", err.Error())
```
Let's also implement the same approach for the below go routine in this function.
Also, if the `runnerErr` becomes unused as a result of this change, let's remove it.

Reply:
done